runtime/lua/vim/lsp.lua

```diff
@@ -887,7 +890,11 @@ function lsp.buf_request(bufnr, method, params, handler)
 
   -- if has client but no clients support the given method, notify the user
   if next(clients) and not method_supported then
-    vim.notify(lsp._unsupported_method(method), vim.log.levels.ERROR)
+    if on_unsupported == nil then
+      vim.notify(lsp._unsupported_method(method), vim.log.levels.ERROR)
+    else
+      on_unsupported()
```

Comment:
Is it not acceptable to throw an error instead (which the caller can choose to handle or not)? Adding random callbacks to provide error handling seems clunky.

Is this another case where we're suffering from the lack of a promise/future abstraction?

Reply:
I also wasn't too keen on adding a new parameter but raising the error would have been a breaking change, and although handling errors with `pcall` works we afaik don't have an established pattern for how to deal with different types of errors. (E.g. if we introduce error codes like some of the uv API does, or if we recommend to parse the error message)

`buf_request` was also already removed from the docs with the intention to replace it eventually, but there wasn't much progress towards a suitable replacement. (Other than the existing `buf_request_all`)

 > Is this another case where we're suffering from the lack of a promise/future abstraction?

I'm not sure if it would help in this case.

Reply:
> `buf_request` was also already removed from the docs with the intention to replace it eventually

Ok that reduces the concern here :)